src/shared/btrfs-util.c

```diff
@@ -1380,8 +1380,24 @@ static int subvol_snapshot_children(
 
         strncpy(vol_args.name, subvolume, sizeof(vol_args.name)-1);
 
-        if (ioctl(new_fd, BTRFS_IOC_SNAP_CREATE_V2, &vol_args) < 0)
-                return -errno;
+        for (;;) {
+                if (ioctl(new_fd, BTRFS_IOC_SNAP_CREATE_V2, &vol_args) < 0)
+                        return -errno;
+
+                if (!FLAGS_SET(flags, BTRFS_SNAPSHOT_LOCK))
+                        break;
+
+                subvolume_fd = xopenat_lock(new_fd, subvolume,
+                                            O_RDONLY|O_NOCTTY|O_CLOEXEC|O_DIRECTORY|O_NOFOLLOW,
+                                            /* xopen_flags = */ 0,
+                                            /* mode = */ 0,
+                                            LOCK_BSD,
+                                            LOCK_EX);
+                if (subvolume_fd >= 0)
```

Comment:
somewhere here there should be some validation that this is actually a subvol that you opened, no?